File: 173-binary-search-tree-iterator/binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    private Stack<TreeNode> stack;

    public BSTIterator(TreeNode root) {
        stack = new Stack<>();
        addLeft(root);
    }
    
    public int next() {
        TreeNode curr = stack.pop();
        if (null!=curr.right) addLeft(curr.right);
        return curr.val;
    }
    
    public boolean hasNext() {
        return !stack.empty();
    }

    private void addLeft(TreeNode root){
        TreeNode temp = root;
        while (null!=temp){
            stack.push(temp);
            temp = temp.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 173-binary-search-tree-iterator/binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        self._addLeft(root)

    def next(self) -> int:
        curr = self.stack.pop()
        if curr.right:
            self._addLeft(curr.right)
        return curr.val

    def hasNext(self) -> bool:
        return self.stack and len(self.stack) > 0

    
    def _addLeft(self, root: Optional[TreeNode]):
        temp = root
        while temp:
            self.stack.append(temp)
            temp = temp.left


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()